iroh-net/src/magic_endpoint.rs

```diff
@@ -0,0 +1,561 @@
+
+        debug!(
+            "connecting to {}: (via {} - {:?})",
+            peer_id, addr, known_addrs
+        );
+
+        // TODO: We'd eventually want to replace "localhost" with something that makes more sense.
+        let connect = self
+            .endpoint
+            .connect_with(client_config, addr, "localhost")?;
+
+        connect.await.context("failed connecting to provider")
+    }
+
+    /// Inform the magic socket about addresses of the peer.
+    ///
+    /// This updates the magic socket's *netmap* with these addresses, which are used as candidates
+    /// when connecting to this peer (in addition to addresses obtained from a derp server).
+    pub async fn add_known_addrs(
+        &self,
+        peer_id: PeerId,
+        addrs: &[SocketAddr],
+    ) -> anyhow::Result<()> {
+        const DEFAULT_DERP_REGION: u16 = 1;
+
+        let node_key: hp::key::node::PublicKey = peer_id.into();
+        let netmap = {
+            let mut netmap = self.netmap.lock().unwrap();
+            let node = netmap.peers.iter_mut().find(|peer| peer.key == node_key);
+            if let Some(node) = node {
+                for addr in addrs {
+                    if !node.endpoints.contains(addr) {
+                        node.endpoints.push(*addr);
+                        node.addresses.push(addr.ip());
+                    }
+                }
+            } else {
+                let endpoints = addrs.to_vec();
+                let addresses = endpoints.iter().map(|ep| ep.ip()).collect();
+                let node = cfg::Node {
+                    name: None,
+                    addresses,
+                    endpoints,
+                    key: node_key.clone(),
+                    derp: Some(DEFAULT_DERP_REGION),
+                };
+                netmap.peers.push(node)
+            }
+            netmap.clone()
+        };
+        self.conn.set_network_map(netmap).await?;
+        Ok(())
+    }
+
+    /// Close the QUIC endpoint and the magic socket.
+    ///
+    /// This will close all open QUIC connections with the provided error_code and reason. See
+    /// [quinn::Connection] for details on how these are interpreted.
+    ///
+    /// It will then wait for all connections to actually be shutdown, and afterwards
+    /// close the magic socket.
+    ///
+    /// Returns an error if closing the magic socket failed.
+    /// TODO: Document error cases.
+    pub async fn close(&self, error_code: VarInt, reason: &[u8]) -> anyhow::Result<()> {
+        self.endpoint.close(error_code, reason);
+        self.endpoint.wait_idle().await;
+        self.conn.close().await?;
+        Ok(())
+    }
+
+    #[cfg(test)]
+    pub(crate) fn conn(&self) -> &Conn {
+        &self.conn
+    }
+    #[cfg(test)]
+    pub(crate) fn endpoint(&self) -> &quinn::Endpoint {
+        &self.endpoint
+    }
+}
+
+/// Accept an incoming connection and extract the client-provided [`PeerId`] and ALPN protocol.
+pub async fn accept_conn(
```

Comment:
I'd be tempted to experiment with an extension trait or a wrapper for `Connecting` (and I guess `Accept`).  But we can do that later.

Reply:
Yup I thought about that too. However we'd either have to use `async_trait` or write the futures by hand. I'd defer this to post-merge I think.